### PR TITLE
#2958 - Part-time NOA Summary Based on e-Cert Values - Part 1

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- The dynamic population of the final awards was kept and the changed code is supposed to follow the same approach already in place, to be clear.
- Part-time final awards are now retrieved from the disbursement effective amount (e-Cert effective values).
  - The controller service now has a specific method to populate part-time and a specific method for full-time (that was already in place and was not changed).
- Date sent added to the result payload for the Ministry and Institution.
- There was a previous check for COE completion that is now replaced for the disbursement status defined as `Sent`. which fits Full-time and Part-time since both need at least the e-Cert generated to have the final awards populated.

### E2E Tests
- Adapted existing E2E tests.
- Added E2E tests for the Ministry (tests were missing for AEST and they were added based in the existing ones for student/institution)
  - Should get the student assessment award summary containing all federal and provincial estimated and final awards for a full-time application with two disbursements when the receipts are available.
  - Should get the student assessment award summary containing estimated and final awards for a part-time application with two disbursements when e-Cert was sent.
  - Should throw not found exception when assessment is not found.
- Add new E2E test for Students
  - Should not generate final award values for a part-time application when the disbursement has not been sent yet.

### Bug fix
- The disbursement `date_sent` was mapped to the wrong DB date type, causing the wrong time to persist in the database.

### Upcoming PR
- UI-related changes.

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+          disbursementReceipt2csgd: 1210,
+          disbursementReceipt2bcag: 1310,
+          disbursementReceipt2sbsd: 1410,
+        },
+      });
+  });
+
+  it("Should throw not found exception when assessment is not found.", async () => {
+    // Arrange
+    const endpoint = `/aest/assessment/99999/award`;
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts:415]: Double quotes.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 2]
```diff
@@ -109,21 +118,23 @@ export class AssessmentControllerService {
    * @param options for NOA.
    * - `includeDocumentNumber` when true document number is mapped
    * to disbursement dynamic data.
+   * - `includeDateSent `  when true date sent is mapped
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:121]: Should have a comma "when true, date sent is mapped"

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 3]
```diff
@@ -256,33 +258,80 @@ export class AssessmentControllerService {
 
   /**
    * Populate the final awards in a dynamic way like disbursement schedule(estimated) awards.
-   * @param disbursementReceipts disbursement receipt details.
+   * @param assessment student assessment.
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:261]: Not part of this ticket but this method doesn't have params `disbursementSchedule` and `identifier`.

**Review by andrepestana-aot — APPROVED** [review]
Great job! Left minor comments.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 4]
```diff
@@ -256,33 +258,80 @@ export class AssessmentControllerService {
...
+    assessment: StudentAssessment,
+    options?: {
+      studentId?: number;
+      applicationId?: number;
+    },
+  ): Promise<DynamicAwardValue | undefined> {
+    const [firstDisbursement] = assessment.disbursementSchedules;
+    if (
+      firstDisbursement.disbursementScheduleStatus !==
+      DisbursementScheduleStatus.Sent
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:280]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 5]
```diff
@@ -256,33 +258,80 @@ export class AssessmentControllerService {
...
+        const awards = await this.populateDisbursementReceiptAwardValues(
+          disbursementReceipts,
+          schedule,
+          `${DISBURSEMENT_RECEIPT_PREFIX}${index++}`,
+        );
+        finalAward = { ...finalAward, ...awards };
       }
+      return finalAward;
+    }
+    // Part-time receipts will not contains all the awards value hence the e-Cert effective
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:309]: will not contain ?

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 6]
```diff
@@ -256,33 +258,80 @@ export class AssessmentControllerService {
...
+
+  /**
+   * Populate final awards values from disbursements receipts.
+   * @param disbursementReceipts disbursement receipts.
+   * @param disbursementSchedule disbursement schedules.
+   * @param identifier identifier which is used to create dynamic data by appending award code to it.
+   * @returns dynamic award data of disbursement receipts of a given disbursement.
    */
-  private populateDisbursementReceiptAwardValues(
+  private async populateDisbursementReceiptAwardValues(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:329]: Why this method is async?

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+    sharedMSFAANumber = await db.msfaaNumber.save(
+      createFakeMSFAANumber(
+        { student: sharedStudent },
+        { msfaaState: MSFAAStates.Signed },
+      ),
+    );
+  });
+
+  it("Should get the student assessment award summary containing all federal and provincial estimated and final awards for a full-time application with two disbursements when the receipts are available.", async () => {
+    const [dateSent1, dateSent2] = [addDays(-30), addDays(-1)];
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts:52]: `//Arrange` missing here.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+          disbursementReceipt2bcsl: 600,
+          disbursementReceipt2bcag: 700,
+          disbursementReceipt2bgpd: 800,
+          disbursementReceipt2sbsd: 900,
+        },
+      });
+  });
+
+  it("Should get the student assessment award summary containing estimated and final awards for a part-time application with two disbursements when e-Cert was sent.", async () => {
+    const [dateSent1, dateSent2] = [addDays(-60), addDays(-5)];
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.aest.controller.getAssessmentAwardDetails.e2e-spec.ts:226]: https://github.com/bcgov/SIMS/pull/3124/files#r1566383755

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts [hunk 9]
```diff
@@ -93,21 +104,19 @@ describe("AssessmentStudentsController(e2e)-getAssessmentAwardDetails", () => {
       {
         offeringIntensity: OfferingIntensity.partTime,
         applicationStatus: ApplicationStatus.Completed,
+        firstDisbursementInitialValues: {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts:107]: Should we add the `date_sent` here to ensure it is not returned by API ??

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:121]: I will add it.

**Comment by dheepak-aot** [comment]
Good job @andrewsignori-aot . Thanks for minor refactors and taking care of data type for `date_sent`.
Left some minor comments.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:261]: I will remove them.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:329]: Back and forth between refactors, I will remove it.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts:107]: I will add it.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 10]
```diff
@@ -361,6 +407,31 @@ export class AssessmentControllerService {
...
   }
 
+  /**
+   * Populate final awards values from e-Cert effective values.
+   * @param disbursementValues disbursement values with the effective amounts used to generate the e-Cert.
+   * @param identifier identifier which is used to create dynamic data by appending award code to it.
+   * @returns dynamic award data of disbursement receipts of a given disbursement.
+   */
+  private populateDisbursementECertAwardValues(
+    disbursementValues: DisbursementValue[],
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:417]: Minor formatting error.

image

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts [hunk 11]
```diff
@@ -521,6 +615,9 @@ describe("AssessmentStudentsController(e2e)-getAssessmentAwardDetails", () => {
       {
         offeringIntensity: OfferingIntensity.fullTime,
         applicationStatus: ApplicationStatus.Completed,
+        firstDisbursementInitialValues: {
+          disbursementScheduleStatus: DisbursementScheduleStatus.Sent,
+        },
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts:620]: date sent is not added for part-time test.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:417]: How, I have no idea. Just did the same and it formatted as expected.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.students.controller.getAssessmentAwardDetails.e2e-spec.ts:620]: I added to one part-time and one full-time.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes. Looks good. 👍

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 12]
```diff
@@ -109,21 +118,23 @@ export class AssessmentControllerService {
...
    */
   private populateDisbursementAwardValues(
     disbursementSchedules: DisbursementSchedule[],
     options?: {
       includeDocumentNumber?: boolean;
+      includeDateSent?: boolean;
       maskMSFAA?: boolean;
     },
-  ): Record<string, string | number> {
+  ): DynamicAwardValue {
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:133]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 2 above)
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:121]: Please remove the extra space after includeDateSent

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts]: Removed

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 13]
```diff
@@ -256,33 +257,78 @@ export class AssessmentControllerService {
...
-   *    disbursementReceipt2bgpd: 1007,
-   *    disbursementReceipt2sbsd: 1008,
+   */
+  private async populateDisbursementFinalAwardsValues(
+    assessment: StudentAssessment,
+    options?: {
+      studentId?: number;
+      applicationId?: number;
+    },
+  ): Promise<DynamicAwardValue | undefined> {
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:272]: why cant we add the undefined in the DynamicAwardValue?

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:272]: We can. Adding it to the possible return type is just to make it explicit.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Pull request merged**